RedBookPlayer/Player.cs

```diff
@@ -453,46 +563,298 @@ public void Rewind()
+            {
+                if(track.TrackSession < _toc.FirstCompleteSession)
+                    _toc.FirstCompleteSession = (byte)track.TrackSession;
 
-        public bool Run = true;
+                if(track.TrackSession <= _toc.LastCompleteSession)
+                {
+                    currentTrack = (byte)track.TrackSequence;
 
-        public PlayerSource(ReadFunction read) => this.read = read;
+                    continue;
+                }
 
-        public WaveFormat WaveFormat => new WaveFormat();
-        bool IAudioSource.CanSeek    => throw new NotImplementedException();
+                if(_toc.LastCompleteSession > 0)
+                    sessionEndingTrack.Add(_toc.LastCompleteSession, currentTrack);
 
-        public long Position
-        {
-            get => throw new NotImplementedException();
-            set => throw new NotImplementedException();
+                _toc.LastCompleteSession = (byte)track.TrackSession;
+            }
+
+            byte currentSession = 0;
+
+            foreach(Track track in Image.Tracks.OrderBy(t => t.TrackSession).ThenBy(t => t.TrackSequence))
+            {
+                _trackFlags.TryGetValue((byte)track.TrackSequence, out byte trackControl);
+
+                if(trackControl == 0 &&
+                   track.TrackType != Aaru.CommonTypes.Enums.TrackType.Audio)
+                    trackControl = (byte)CdFlags.DataTrack;
+
+                // Lead-Out
+                if(track.TrackSession > currentSession &&
+                   currentSession != 0)
+                {
+                    (byte minute, byte second, byte frame) leadoutAmsf = LbaToMsf(track.TrackStartSector - 150);
+
+                    (byte minute, byte second, byte frame) leadoutPmsf =
+                        LbaToMsf(Image.Tracks.OrderBy(t => t.TrackSession).ThenBy(t => t.TrackSequence).Last().
+                                        TrackStartSector);
+
+                    // Lead-out
+                    trackDescriptors.Add(new TrackDataDescriptor
+                    {
+                        SessionNumber = currentSession,
+                        POINT = 0xB0,
+                        ADR = 5,
+                        CONTROL = 0,
+                        HOUR = 0,
+                        Min = leadoutAmsf.minute,
+                        Sec = leadoutAmsf.second,
+                        Frame = leadoutAmsf.frame,
+                        PHOUR = 2,
+                        PMIN = leadoutPmsf.minute,
+                        PSEC = leadoutPmsf.second,
+                        PFRAME = leadoutPmsf.frame
+                    });
+
+                    // This seems to be constant? It should not exist on CD-ROM but CloneCD creates them anyway
+                    // Format seems like ATIP, but ATIP should not be as 0xC0 in TOC...
+                    trackDescriptors.Add(new TrackDataDescriptor
```

Comment:
You don't need C0 point.

Reply:
This entire block was copied whole from CloneCd, so I'll admit I don't fully understand how it all works. I'll comment out that block.